shared/src/balancer_event_handler.rs

```diff
@@ -3,195 +3,140 @@ use crate::{
+#[derive(Derivative)]
+#[derivative(Debug)]
 pub struct BalancerPools {
     /// Used for O(1) access to all pool_ids for a given token
     pools_by_token: HashMap<H160, HashSet<H256>>,
     /// WeightedPool data for a given PoolId
     pools: HashMap<H256, WeightedPool>,
-    /// Temporary storage for WeightedPools containing insufficient constructor data
-    pending_pools: HashMap<H256, WeightedPoolBuilder>,
-}
-
-/// There are three specialization settings for Pools,
-/// which allow for cheaper swaps at the cost of reduced functionality:
-#[derive(Debug, Clone, Copy, Eq, Hash, PartialEq)]
-#[repr(u8)]
-pub enum PoolSpecialization {
-    /// no specialization, suited for all Pools. IGeneralPool is used for swap request callbacks,
-    /// passing the balance of all tokens in the Pool. These Pools have the largest swap costs
-    /// (because of the extra storage reads), which increase with the number of registered tokens.
-    General = 0,
-    /// IMinimalSwapInfoPool is used instead of IGeneralPool, which saves gas by only passing the
-    /// balance of the two tokens involved in the swap. This is suitable for some pricing algorithms,
-    /// like the weighted constant product one popularized by Balancer V1. Swap costs are
-    /// smaller compared to general Pools, and are independent of the number of registered tokens.
-    MinimalSwapInfo = 1,
-    /// only allows two tokens to be registered. This achieves the lowest possible swap gas cost.
-    /// Like minimal swap info Pools, these are called via IMinimalSwapInfoPool.
-    TwoToken = 2,
-}
-
-impl PoolSpecialization {
-    fn new(specialization: u8) -> Result<Self> {
-        match specialization {
-            0 => Ok(Self::General),
-            1 => Ok(Self::MinimalSwapInfo),
-            2 => Ok(Self::TwoToken),
-            t => Err(anyhow!("Invalid PoolSpecialization value {} (> 2)", t)),
-        }
-    }
+    #[derivative(Debug = "ignore")]
+    data_fetcher: Box<dyn PoolDataFetching>,
 }
 
 impl BalancerPools {
-    fn try_upgrade(&mut self) {
-        for (pool_id, pool_builder) in self.pending_pools.clone() {
-            if let Some(weighted_pool) = pool_builder.into_pool() {
-                // When upgradable, delete pending pool and add to valid pools
-                tracing::info!("Upgrading Pool Builder with id {:?}", pool_id);
-                self.pools.insert(pool_id, weighted_pool.clone());
-                self.pending_pools.remove(&pool_id);
-                for token in weighted_pool.tokens {
-                    self.pools_by_token
-                        .entry(token)
-                        .or_default()
-                        .insert(pool_id);
-                }
+    async fn insert_events(&mut self, events: Vec<(EventIndex, PoolCreated)>) -> Result<()> {
+        for (index, creation) in events {
+            let weighted_pool =
+                WeightedPool::from_event(index.block_number, creation, &*self.data_fetcher).await?;
```

Comment:
In a follow up PR I would propose that these be parallelized, but I would like to keep the changes here contained to the storing of the weighted pools.

Reply:
After performing some naive benchmarks, I see no reason to believe that our data fetching is or will ever be a bottle neck even when it is not batched or parallelized. In fact, from what we see here (of course this is only data from a product which is not yet live), there would be no improvement gained from batched our requests. Only when several pools are deployed within the same 100 (25 minutes w. 15 second block times) blocks can we expect to see any marginal performance gain from batching requests.

I have found from a single simulation contained in the data logs below that fetching token data for 2 pools and syncing events for about 500 blocks takes < 0.4 seconds (including all other processes that the orderbook API is running in parallel). Only when several pools are deployed within the same 100 blocks (25 minutes w. 15 second block times) can we expect to see any marginal performance gain from batching requests. 

Say, for example that 100 pools are deployed in 25 minutes... then it is expected to take 20 seconds to process and store these events (but of course this would happen without notice at different times during that 25 minute interval).

## Summary of finding from logs:

Spans over:
    1.5 months (April 21 to June 9) - 316_781 blocks

Total time to sync all pools: 8 minutes.

## Points of interest:
- There is only 2 occurrences of > 1 pool created over a period of 1000 block (and these are the real time gaps).
- The real time gap here is not with the pool data fetching, but merely with event sync
      - Between 11 and 12 (128_443 blocks) we have a time gap of 2 minutes.
      - Between 1 and 3: total time gap of 0.4 seconds to fetch the pool info for 2 new pools.

## Comparison with Uniswap V2:

- There are 41_928 pools total. (dune query: `select count(*) from uniswap_v2."Factory_evt_PairCreated"`)
- Total processing time in terms of data fetching (plus 500 block sync) 41_928 * 0.2 (seconds) = 2.3 hours
- For a cold start, this is not ideal!
- However, we can easily handle 5000 pools in 16 minutes.
By this point, I would imagine that we can read/write to an from file to to avoid such issues.

Of course, when comparing with UniswapV2, this looks terrible (especially for a cold start). However, by the time BalancerV2 reaches this magnitude, it should be fairly straightforward to feed a backup of this data into the program at runtime.

## File Size: 
The current amount of data when stored as JSON (without white space or line breaks) is 18kb (2kb compressed). For 20 pools, this makes an average of 1kb per pool. The file store size is then linear with the number of pools. So even when/if balancer reaches the magnitude of uniswapV2 this file should not exceed 42 mb (or 4.2 mb compressed). See file [BalancerPools.txt](https://github.com/gnosis/gp-v2-services/files/6634167/BalancerPools.txt) and compressed [BalancerPools.zip](https://github.com/gnosis/gp-v2-services/files/6634178/BalancerPools.zip)


<details><summary>raw data from naive benchmark</summary>
2021-06-10T19:29:39.822Z Last event block 12285749 - Balancer Pool State: 1  pools in store
2021-06-10T19:29:40.201Z Last event block 12286276 - Balancer Pool State: 3  pools in store
2021-06-10T19:29:57.191Z Last event block 12297008 - Balancer Pool State: 4  pools in store
2021-06-10T19:30:00.055Z Last event block 12299382 - Balancer Pool State: 5  pools in store
2021-06-10T19:30:05.540Z Last event block 12303593 - Balancer Pool State: 6  pools in store
2021-06-10T19:30:06.723Z Last event block 12304403 - Balancer Pool State: 7  pools in store
2021-06-10T19:30:23.915Z Last event block 12318074 - Balancer Pool State: 8  pools in store
2021-06-10T19:30:49.069Z Last event block 12331542 - Balancer Pool State: 9  pools in store
2021-06-10T19:30:54.249Z Last event block 12331765 - Balancer Pool State: 10 pools in store
2021-06-10T19:32:16.895Z Last event block 12376911 - Balancer Pool State: 11 pools in store
2021-06-10T19:34:51.197Z Last event block 12505354 - Balancer Pool State: 12 pools in store
2021-06-10T19:34:54.090Z Last event block 12509636 - Balancer Pool State: 13 pools in store
2021-06-10T19:36:11.955Z Last event block 12563369 - Balancer Pool State: 14 pools in store
2021-06-10T19:36:27.052Z Last event block 12567967 - Balancer Pool State: 15 pools in store
2021-06-10T19:37:14.006Z Last event block 12595481 - Balancer Pool State: 16 pools in store
2021-06-10T19:37:15.167Z Last event block 12595681 - Balancer Pool State: 17 pools in store
2021-06-10T19:37:25.202Z Last event block 12602530 - Balancer Pool State: 20 pools in store
</details>